File: server/csharp/MetaObjects.Render/Provider.cs
// A provider resolves a 2-layer logical reference (`group/source`) to template
// text. The render engine never does I/O itself — it delegates resolution to
// the injected provider, so a fixed provider makes Render() deterministic.
//
// Ported from typescript/packages/render/src/provider.ts.

namespace MetaObjects.Render;

/// <summary>Resolves a <c>group/source</c> reference to its text, or null if absent.</summary>
public interface IProvider
{
    /// <summary>Resolve a <c>group/source</c> reference to its text, or null if absent.</summary>
    string? Resolve(string reference);
}

/// <summary>Deterministic, in-memory provider (the conformance + test provider).</summary>
public sealed class InMemoryProvider(IReadOnlyDictionary<string, string> map) : IProvider
{
    private readonly IReadOnlyDictionary<string, string> _map = map;

    public string? Resolve(string reference) =>
        _map.TryGetValue(reference, out var text) ? text : null;
}

File: server/typescript/packages/render/src/provider.ts
// A provider resolves a 2-layer logical reference (`group/source`) to template
// text. The render engine never does I/O itself — it delegates resolution to
// the injected provider, so a fixed provider makes render() deterministic.

export interface Provider {
  /** Resolve a `group/source` reference to its text, or undefined if absent. */
  resolve(ref: string): string | undefined;
}

/** Deterministic, in-memory provider (the conformance + test provider). */
export class InMemoryProvider implements Provider {
  constructor(private readonly map: Record<string, string>) {}
  resolve(ref: string): string | undefined {
    return Object.prototype.hasOwnProperty.call(this.map, ref) ? this.map[ref] : undefined;
  }
}
